### PR TITLE
BNAS: Allow disabling BN Adaptation

### PR DESCRIPTION
**Pull request by jpablomch**
### Changes

(1) When num_bn_adaptation_samples = 0, batchnorm adaptation will be skipped
(2) Add function "get_last_lr". This function will be used in Hugging Face patch.
(3) Add property for lr_scheduler. This will be used in Hugging Face patch.

### Reason for changes

Substep to implement a patch for HF Transformers for BootstrapNAS

### Related tickets

N/A

### Tests

N/A

### CODE REVIEW

File: nncf/experimental/torch/nas/bootstrapNAS/search/search.py [hunk 1]
```diff
@@ -341,15 +343,19 @@ def get_macs_for_active_subnet() -> float:
 
         if self.best_config is not None:
             self._elasticity_ctrl.multi_elasticity_handler.activate_subnet_for_config(self.best_config)
-            self.bn_adaptation.run(self._model)
+            if self.bn_adaptation is not None:
+                self.bn_adaptation.run(self._model)
```
Comment (ljaljushkin) [inline nncf/experimental/torch/nas/bootstrapNAS/search/search.py:347]: This part doesn't seem working. Because Search Algorithm always instantiates `bn_adaptation`: https://github.com/openvinotoolkit/nncf/pull/1329/files#diff-7afaa162c96c4fe46d2bce9b80be7866d27d18e0914da3794e31cbabf007ff96R218-R220

Please add unit test checking that bn_adaptation is called or not called depending on num_samples provided in config. 
Here's the pseudocode:
```python
@pytest.mark.parametrize( ('bn_adapt_section', 'is_called'), 
(
    (section_with_zero_num_samples, False), 
    (section_with_non_zero_num_samples, True)
)
def test_bn_adapt(mocker, bn_adapt_section, is_called)
     nncf_config = NNCFConfig(...)
     update_bn_adapt_section(nncf_config, bn_adapt_section) # to be implemented
     mock_nncf_network = mocker.MagicMock(spec=NNCFNetwork)
     mock_ctrl =mocker..MagicMock(spec=ElasticityController)
     bn_adapt_run_patch = 
     mocker.patch('nncf.common.initialization.batchnorm_adaptation.BatchnormAdaptationAlgorithm.run')
     search_algo = SearchAlgorithm(mock_nncf_network, mock_ctrl, nncf_config)
     search_algo.run()
     if is_called:
          bn_adapt_run_patch.assert_called_once()
     else:
          bn_adapt_run_patch.assert_not_called()
```

File: nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def __init__(self, nncf_config: NNCFConfig, should_init: bool = True):
         super().__init__(nncf_config, should_init)
         bn_adapt_params = self._algo_config.get('batchnorm_adaptation', {})
         bn_adapt_algo_kwargs = get_bn_adapt_algo_kwargs(nncf_config, bn_adapt_params)
-        self._bn_adaptation = BatchnormAdaptationAlgorithm(**bn_adapt_algo_kwargs)
+        self._bn_adaptation = BatchnormAdaptationAlgorithm(**bn_adapt_algo_kwargs) if bn_adapt_algo_kwargs else None
```
Comment (ljaljushkin) [inline nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py:51]: Again, this logic would break Training. There's lack of unit testing of case when `bn_adaptation` is disabled.

Currently, training will fail if bn_adaptation is empty, which would be not expected behavior.
```
    def _run_batchnorm_adaptation(self, model):
        if self._bn_adaptation is None:
            raise RuntimeError("Missing initialization of Batchnorm Adaptation algorithm")
        self._bn_adaptation.run(model)
```
Propose to verify behavior of Progressive Shrinking Controller in case of different `num_samples` for BNAdaptation.
Here's an example of unit test for Progressive Shrinking Controller : 
https://github.com/openvinotoolkit/nncf/blob/develop/tests/torch/nas/test_scheduler.py#L125
The test structure can be the same as it was suggested for SearchAlgorithm above.

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/experimental/torch/nas/bootstrapNAS/search/search.py [hunk 3]
```diff
@@ -331,6 +331,8 @@ def get_macs_for_active_subnet() -> float:
                                                                             self._elasticity_ctrl)
         self._evaluator_handlers.append(self._efficiency_evaluator_handler)
         self._evaluator_handlers.append(self._accuracy_evaluator_handler)
+        self.maximal_vals = [evaluator_handler.current_value for evaluator_handler
```
[the remark this answers is not in the document]
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/search/search.py:334]: Thanks! Moved activation of maximal subnet to `run`.

File: nncf/experimental/torch/nas/bootstrapNAS/search/search.py (hunk 1 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/search/search.py:347]: Thanks! Added test.

File: nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py (hunk 2 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py:51]: Thanks! Added test similar to search. 
In https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py#L50, we check Batchnorm adapt and number of samples. 
In ProgressiveShrinkingController we check if batchnorm adapt is enabled when preparing for validation and when setting stage, preventing to run _run_batchnorm_adaptation(..) if bn adapt is not enabled. 

We warn the user if _run_batchnorm_adaptation(..) is called directly, as well.

File: tests/torch/nas/test_ps_controller.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        })
+        mock_handler = MultiElasticityHandler(handlers, mock_nncf_network)
+        # pylint:disable=protected-access
+        mock_elasticity_ctrl = mocker.stub()
+        mock_elasticity_ctrl.multi_elasticity_handler = mock_handler
+        lr_schedule_config = {}
+        training_config = nncf_config.get('bootstrapNAS', {}).get('training', {})
+        bn_adapt_params = training_config.get('batchnorm_adaptation', {})
+        bn_adapt_algo_kwargs = get_bn_adapt_algo_kwargs(nncf_config, bn_adapt_params)
+        bn_adaptation = BatchnormAdaptationAlgorithm(**bn_adapt_algo_kwargs) if bn_adapt_algo_kwargs else None
```
Comment (ljaljushkin) [inline tests/torch/nas/test_ps_controller.py:118]: The whole point of the behavior test is lost here. 
Instead of repeating the handling `bn_adapt`, you need to verify that it's correctly handled in the training algorithm.

Propose for the training to repeat user scenario.
Just create training algorithm with mocked nncf_network and with patched run method of `BatchnormAdaptationAlgorithm`:
```python
 bn_adapt_run_patch = mocker.patch(
            "nncf.common.initialization.batchnorm_adaptation.BatchnormAdaptationAlgorithm.run")
training_algorithm = EpochBasedTrainingAlgorithm.from_checkpoint(nncf_network, bn_adapt_args,
                                                                         resuming_checkpoint_path)
```

File: tests/torch/nas/test_search.py [hunk 5]
```diff
@@ -160,6 +169,29 @@ def test_design_upper_bounds(self, nas_model_name):
         assert search.vars_upper == NAS_MODELS_SEARCH_ENCODING[nas_model_name]
         assert search.num_vars == len(NAS_MODELS_SEARCH_ENCODING[nas_model_name])
 
+    @pytest.mark.parametrize(('bn_adapt_section', 'is_called'), (("section_with_zero_num_samples", False),
+                                                                 ("section_with_non_zero_num_samples", True)))
+    def test_bn_adapt(self, mocker, bn_adapt_section, is_called, tmp_path):
```
Comment (ljaljushkin) [inline tests/torch/nas/test_search.py:174]: `bn_adapt_section` is not needed

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/torch/nas/test_search.py (hunk 5 above)
Comment (jpablomch) [reply tests/torch/nas/test_search.py]: Fixed. Thanks!

File: tests/torch/nas/test_ps_controller.py (hunk 4 above)
Comment (jpablomch) [reply tests/torch/nas/test_ps_controller.py]: Thanks! Using EpochBasedTrainingAlgorithm. We also modified builder state because it wasn't saving bn adapt parameters. When we load state with "EpochBasedTrainingAlgorithm.from_checkpoint" it was using default 'num_bn_adaptation_samples' (i.e. 2000). Thanks!

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/torch/nas/test_ps_controller.py (hunk 4 above)
Comment (ljaljushkin) [reply tests/torch/nas/test_ps_controller.py]: 👍

File: tests/torch/nas/test_ps_controller.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            EBTrainAlgoStateNames.MIN_SUBNET_BEST_ACC1: 0,
+            EBTrainAlgoStateNames.OPTIMIZER: None,
+            EBTrainAlgoStateNames.TRAINING_ALGO_STATE: ctrl.get_compression_state()
+        }
+        torch.save(checkpoint, checkpoint_path)
+
+        bn_adapt_run_patch = mocker.patch(
+            "nncf.common.initialization.batchnorm_adaptation.BatchnormAdaptationAlgorithm.run")
+        clean_model = create_nncf_network(clean_model, nncf_config)
+        training_algorithm = EpochBasedTrainingAlgorithm.from_checkpoint(clean_model, bn_adapt_args, checkpoint_path)
```
[the remark this answers is not in the document]
Comment (jpablomch) [reply tests/torch/nas/test_ps_controller.py]: Thanks! Modified the test to using config instead of checkpoint.

File: tests/torch/nas/test_all_elasticity.py [hunk 7]
```diff
@@ -397,7 +399,10 @@ def test_can_restore_from_state():
     _, training_ctrl = resume_compression_from_state(nncf_network, old_state, empty_nncf_config)
 
     new_state = training_ctrl.get_compression_state()
-    assert new_state == old_state
+    assert old_state == REF_COMPRESSION_STATE_FOR_TWO_CONV
+    REF_NEW_STATE = deepcopy(REF_COMPRESSION_STATE_FOR_TWO_CONV)
+    REF_NEW_STATE['builder_state']['progressive_shrinking'].update({'bn_adaptation_params': {}})
```
Comment (ljaljushkin) [inline tests/torch/nas/test_all_elasticity.py:404]: why do you need this? why are `bn_adapt_params` not saved as before?
according to the code, it should be saved:
https://github.com/openvinotoolkit/nncf/pull/1329/files#diff-80b9e15adbbbec089ebbe4a96e83fac57b8434bc086d5404ce2cd084e896a7b1R110

Comment (jpablomch) [reply tests/torch/nas/test_all_elasticity.py]: Thanks! I have reverted to the previous form of the test, and fixed how bn_params are handled.

**Review by ljaljushkin — APPROVED** [review]

**Comment by ljaljushkin** [comment]
run pytorch pre-commit tests

**Comment by ljaljushkin** [comment]
please fix tests
image

**Comment by jpablomch** [comment]
run build
